File: packages/mpc-wallet-sdk/src/chains/evm.ts
/**
 * EVM chain adapter
 */

import type { Balance, Signature, TxHash, TxReceipt } from '../types';

/**
 * EVM chain configuration
 */
export interface EvmChainConfig {
  /** Chain ID */
  chainId: number;
  /** Chain name */
  name: string;
  /** RPC URLs (with failover) */
  rpcUrls: string[];
  /** Native currency symbol */
  symbol: string;
  /** Native currency decimals */
  decimals: number;
  /** Block explorer URL */
  explorerUrl?: string;
}

/**
 * Pre-configured chain configs
 */
export const EVMChains = {
  ETHEREUM_MAINNET: {
    chainId: 1,
    name: 'Ethereum Mainnet',
    rpcUrls: ['https://eth.llamarpc.com', 'https://rpc.ankr.com/eth'],
    symbol: 'ETH',
    decimals: 18,
    explorerUrl: 'https://etherscan.io',
  },
  ETHEREUM_SEPOLIA: {
    chainId: 11155111,
    name: 'Ethereum Sepolia',
    rpcUrls: ['https://sepolia.drpc.org', 'https://rpc.ankr.com/eth_sepolia'],
    symbol: 'ETH',
    decimals: 18,
    explorerUrl: 'https://sepolia.etherscan.io',
  },
  ARBITRUM_ONE: {
    chainId: 42161,
    name: 'Arbitrum One',
    rpcUrls: ['https://arb1.arbitrum.io/rpc', 'https://rpc.ankr.com/arbitrum'],
    symbol: 'ETH',
    decimals: 18,
    explorerUrl: 'https://arbiscan.io',
  },
  OPTIMISM: {
    chainId: 10,
    name: 'Optimism',
    rpcUrls: ['https://mainnet.optimism.io', 'https://rpc.ankr.com/optimism'],
    symbol: 'ETH',
    decimals: 18,
    explorerUrl: 'https://optimistic.etherscan.io',
  },
  BASE: {
    chainId: 8453,
    name: 'Base',
    rpcUrls: ['https://mainnet.base.org', 'https://base.llamarpc.com'],
    symbol: 'ETH',
    decimals: 18,
    explorerUrl: 'https://basescan.org',
  },
  POLYGON: {
    chainId: 137,
    name: 'Polygon',
    rpcUrls: [
      'https://polygon-rpc.com',
      'https://rpc.ankr.com/polygon',
    ],
    symbol: 'MATIC',
    decimals: 18,
    explorerUrl: 'https://polygonscan.com',
  },
} as const;

/**
 * Transaction parameters for EVM
 */
export interface EvmTxParams {
  /** Sender address */
  from: string;
  /** Recipient address */
  to: string;
  /** Value in wei (as string) */
  value: string;
  /** Transaction data (hex) */
  data?: string;
  /** Gas limit */
  gasLimit?: bigint;
  /** Max fee per gas (EIP-1559) */
  maxFeePerGas?: bigint;
  /** Max priority fee per gas (EIP-1559) */
  maxPriorityFeePerGas?: bigint;
  /** Nonce override */
  nonce?: number;
}

/**
 * Unsigned EVM transaction
 */
export interface UnsignedEvmTx {
  /** Chain ID */
  chainId: number;
  /** Serialized transaction for signing */
  serialized: Uint8Array;
  /** Transaction hash for signing */
  signingHash: Uint8Array;
  /** Human-readable summary */
  summary: {
    from: string;
    to: string;
    value: string;
    estimatedFee: string;
  };
}

/**
 * EVM chain adapter
 *
 * @example
 * ```typescript
 * const adapter = new EvmAdapter(EVMChains.ETHEREUM_MAINNET);
 *
 * // Get balance
 * const balance = await adapter.getBalance('0x...');
 *
 * // Build transaction
 * const unsignedTx = await adapter.buildTransaction({
 *   from: '0x...',
 *   to: '0x...',
 *   value: '1000000000000000000', // 1 ETH
 * });
 *
 * // After MPC signing, broadcast
 * const txHash = await adapter.broadcast(signedTx);
 * ```
 */
export class EvmAdapter {
  private config: EvmChainConfig;
  private currentRpcIndex = 0;

  constructor(config: EvmChainConfig) {
    this.config = config;
  }

  /** Get chain ID */
  get chainId(): number {
    return this.config.chainId;
  }

  /** Get native currency symbol */
  get symbol(): string {
    return this.config.symbol;
  }

  /** Get native currency decimals */
  get decimals(): number {
    return this.config.decimals;
  }

  /**
   * Get balance for an address
   */
  async getBalance(address: string): Promise<Balance> {
    const result = await this.rpcCall<string>('eth_getBalance', [
      address,
      'latest',
    ]);

    const rawValue = BigInt(result);
    const formatted = this.formatValue(rawValue);

    return {
      raw: rawValue.toString(),
      formatted,
      symbol: this.config.symbol,
      decimals: this.config.decimals,
    };
  }

  /**
   * Get nonce for an address
   */
  async getNonce(address: string): Promise<number> {
    const result = await this.rpcCall<string>('eth_getTransactionCount', [
      address,
      'latest',
    ]);
    return parseInt(result, 16);
  }

  /**
   * Get current gas prices (EIP-1559)
   */
  async getGasPrices(): Promise<{
    baseFee: bigint;
    maxFeePerGas: bigint;
    maxPriorityFeePerGas: bigint;
  }> {
    // Get latest block for base fee
    const block = await this.rpcCall<{ baseFeePerGas: string }>(
      'eth_getBlockByNumber',
      ['latest', false]
    );

    const baseFee = BigInt(block.baseFeePerGas || '0');

    // Get priority fee suggestion
    const priorityFee = await this.rpcCall<string>(
      'eth_maxPriorityFeePerGas',
      []
    ).catch(() => '0x3b9aca00'); // Default to 1 gwei

    const maxPriorityFeePerGas = BigInt(priorityFee);
    const maxFeePerGas = baseFee * 2n + maxPriorityFeePerGas;

    return {
      baseFee,
      maxFeePerGas,
      maxPriorityFeePerGas,
    };
  }

  /**
   * Estimate gas for a transaction
   */
  async estimateGas(params: EvmTxParams): Promise<bigint> {
    const result = await this.rpcCall<string>('eth_estimateGas', [
      {
        from: params.from,
        to: params.to,
        value: '0x' + BigInt(params.value).toString(16),
        data: params.data || '0x',
      },
    ]);
    return BigInt(result);
  }

  /**
   * Build an unsigned transaction
   */
  async buildTransaction(params: EvmTxParams): Promise<UnsignedEvmTx> {
    // Get nonce if not provided
    const nonce = params.nonce ?? (await this.getNonce(params.from));

    // Get gas prices
    const gasPrices = await this.getGasPrices();
    const maxFeePerGas = params.maxFeePerGas ?? gasPrices.maxFeePerGas;
    const maxPriorityFeePerGas =
      params.maxPriorityFeePerGas ?? gasPrices.maxPriorityFeePerGas;

    // Estimate gas if not provided
    const gasLimit =
      params.gasLimit ?? ((await this.estimateGas(params)) * 120n) / 100n; // 20% buffer

    // Build EIP-1559 transaction
    const tx = {
      chainId: this.config.chainId,
      nonce,
      maxFeePerGas,
      maxPriorityFeePerGas,
      gasLimit,
      to: params.to,
      value: BigInt(params.value),
      data: params.data ? hexToBytes(params.data) : new Uint8Array(0),
    };

    // Serialize transaction (simplified - in production use RLP encoding)
    const serialized = new TextEncoder().encode(JSON.stringify(tx));

    // Compute signing hash (simplified - in production use proper EIP-1559 hash)
    const signingHash = await sha256(serialized);

    // Estimate fee
    const estimatedFee = gasLimit * maxFeePerGas;

    return {
      chainId: this.config.chainId,
      serialized,
      signingHash,
      summary: {
        from: params.from,
        to: params.to,
        value: this.formatValue(BigInt(params.value)),
        estimatedFee: this.formatValue(estimatedFee),
      },
    };
  }

  /**
   * Finalize a transaction with signature
   */
  finalizeTransaction(
    unsignedTx: UnsignedEvmTx,
    signature: Signature
  ): Uint8Array {
    // Combine transaction with signature
    // In production, this would use proper RLP encoding
    const r = hexToBytes(signature.r);
    const s = hexToBytes(signature.s);

    // EIP-1559: v = yParity (0 or 1)
    const v = signature.recoveryId;

    const signed = new Uint8Array(unsignedTx.serialized.length + 65);
    signed.set(unsignedTx.serialized);
    signed.set(r, unsignedTx.serialized.length);
    signed.set(s, unsignedTx.serialized.length + 32);
    signed[unsignedTx.serialized.length + 64] = v;

    return signed;
  }

  /**
   * Broadcast a signed transaction
   */
  async broadcast(signedTx: Uint8Array): Promise<TxHash> {
    const txHex = '0x' + bytesToHex(signedTx);
    const result = await this.rpcCall<string>('eth_sendRawTransaction', [txHex]);

    return {
      hash: result,
      explorerUrl: this.config.explorerUrl
        ? `${this.config.explorerUrl}/tx/${result}`
        : undefined,
    };
  }

  /**
   * Wait for transaction confirmation
   */
  async waitForConfirmation(
    txHash: string,
    timeoutMs = 60000
  ): Promise<TxReceipt> {
    const startTime = Date.now();

    while (Date.now() - startTime < timeoutMs) {
      try {
        const receipt = await this.rpcCall<{
          blockNumber: string;
          status: string;
          gasUsed: string;
          effectiveGasPrice: string;
        } | null>('eth_getTransactionReceipt', [txHash]);

        if (receipt) {
          return {
            txHash,
            blockNumber: parseInt(receipt.blockNumber, 16),
            status: receipt.status === '0x1' ? 'success' : 'failed',
            gasUsed: BigInt(receipt.gasUsed),
            effectiveGasPrice: BigInt(receipt.effectiveGasPrice),
          };
        }
      } catch {
        // Transaction not found yet
      }

      await sleep(2000);
    }

    throw new Error(`Transaction ${txHash} not confirmed within timeout`);
  }

  /**
   * Derive address from public key
   */
  deriveAddress(publicKey: Uint8Array): string {
    // Keccak256 of uncompressed public key (skip first byte if compressed)
    const pk = publicKey[0] === 0x04 ? publicKey.slice(1) : publicKey;
    // In production, use proper keccak256
    const hash = new Uint8Array(32);
    for (let i = 0; i < pk.length; i++) {
      hash[i % 32] ^= pk[i];
    }
    return '0x' + bytesToHex(hash.slice(12));
  }

  /**
   * Check if an address is valid
   */
  isValidAddress(address: string): boolean {
    return /^0x[0-9a-fA-F]{40}$/.test(address);
  }

  /**
   * Get explorer URL for a transaction
   */
  getExplorerTxUrl(txHash: string): string | undefined {
    if (!this.config.explorerUrl) return undefined;
    return `${this.config.explorerUrl}/tx/${txHash}`;
  }

  /**
   * Get explorer URL for an address
   */
  getExplorerAddressUrl(address: string): string | undefined {
    if (!this.config.explorerUrl) return undefined;
    return `${this.config.explorerUrl}/address/${address}`;
  }

  // ============================================================================
  // Private Methods
  // ============================================================================

  private async rpcCall<T>(method: string, params: unknown[]): Promise<T> {
    const errors: Error[] = [];

    for (let i = 0; i < this.config.rpcUrls.length; i++) {
      const rpcUrl = this.config.rpcUrls[this.currentRpcIndex];

      try {
        const response = await fetch(rpcUrl, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            jsonrpc: '2.0',
            method,
            params,
            id: Date.now(),
          }),
        });

        const data = await response.json();

        if (data.error) {
          throw new Error(data.error.message || 'RPC error');
        }

        return data.result;
      } catch (error) {
        errors.push(error as Error);
        this.currentRpcIndex =
          (this.currentRpcIndex + 1) % this.config.rpcUrls.length;
      }
    }

    throw new Error(
      `All RPC endpoints failed: ${errors.map((e) => e.message).join(', ')}`
    );
  }

  private formatValue(value: bigint): string {
    const divisor = 10n ** BigInt(this.config.decimals);
    const whole = value / divisor;
    const fraction = value % divisor;

    if (fraction === 0n) {
      return `${whole} ${this.config.symbol}`;
    }

    const fractionStr = fraction.toString().padStart(this.config.decimals, '0');
    const trimmed = fractionStr.replace(/0+$/, '');
    return `${whole}.${trimmed} ${this.config.symbol}`;
  }
}

// ============================================================================
// Helper Functions
// ============================================================================

function hexToBytes(hex: string): Uint8Array {
  const cleanHex = hex.startsWith('0x') ? hex.slice(2) : hex;
  const bytes = new Uint8Array(cleanHex.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(cleanHex.substr(i * 2, 2), 16);
  }
  return bytes;
}

function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}

async function sha256(data: Uint8Array): Promise<Uint8Array> {
  if (typeof crypto !== 'undefined' && crypto.subtle) {
    const hash = await crypto.subtle.digest('SHA-256', data.buffer as ArrayBuffer);
    return new Uint8Array(hash);
  }
  // Fallback simple hash
  const result = new Uint8Array(32);
  for (let i = 0; i < data.length; i++) {
    result[i % 32] ^= data[i];
  }
  return result;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: packages/mpc-wallet-python/src/mpc_wallet/chains/evm.py
"""EVM chain adapter."""

from dataclasses import dataclass, field
from typing import Any
import asyncio
import httpx

from ..types import Balance, Signature, TxHash, TxReceipt


@dataclass
class EvmChainConfig:
    """EVM chain configuration."""

    chain_id: int
    name: str
    rpc_urls: list[str]
    symbol: str
    decimals: int = 18
    explorer_url: str | None = None


# Pre-configured chain configs
EVMChains = {
    "ETHEREUM_MAINNET": EvmChainConfig(
        chain_id=1,
        name="Ethereum Mainnet",
        rpc_urls=["https://eth.llamarpc.com", "https://rpc.ankr.com/eth"],
        symbol="ETH",
        explorer_url="https://etherscan.io",
    ),
    "ETHEREUM_SEPOLIA": EvmChainConfig(
        chain_id=11155111,
        name="Ethereum Sepolia",
        rpc_urls=["https://sepolia.drpc.org", "https://rpc.ankr.com/eth_sepolia"],
        symbol="ETH",
        explorer_url="https://sepolia.etherscan.io",
    ),
    "ARBITRUM_ONE": EvmChainConfig(
        chain_id=42161,
        name="Arbitrum One",
        rpc_urls=["https://arb1.arbitrum.io/rpc", "https://rpc.ankr.com/arbitrum"],
        symbol="ETH",
        explorer_url="https://arbiscan.io",
    ),
    "OPTIMISM": EvmChainConfig(
        chain_id=10,
        name="Optimism",
        rpc_urls=["https://mainnet.optimism.io", "https://rpc.ankr.com/optimism"],
        symbol="ETH",
        explorer_url="https://optimistic.etherscan.io",
    ),
    "BASE": EvmChainConfig(
        chain_id=8453,
        name="Base",
        rpc_urls=["https://mainnet.base.org", "https://base.llamarpc.com"],
        symbol="ETH",
        explorer_url="https://basescan.org",
    ),
    "POLYGON": EvmChainConfig(
        chain_id=137,
        name="Polygon",
        rpc_urls=["https://polygon-rpc.com", "https://rpc.ankr.com/polygon"],
        symbol="MATIC",
        explorer_url="https://polygonscan.com",
    ),
}


@dataclass
class EvmTxParams:
    """Transaction parameters for EVM."""

    from_address: str
    to: str
    value: str  # Value in wei as string
    data: str | None = None
    gas_limit: int | None = None
    max_fee_per_gas: int | None = None
    max_priority_fee_per_gas: int | None = None
    nonce: int | None = None


@dataclass
class UnsignedEvmTx:
    """Unsigned EVM transaction."""

    chain_id: int
    serialized: bytes
    signing_hash: bytes
    summary: dict[str, str]


class EvmAdapter:
    """
    EVM chain adapter.

    Example:
        >>> adapter = EvmAdapter(EVMChains["ETHEREUM_MAINNET"])
        >>>
        >>> # Get balance
        >>> balance = await adapter.get_balance("0x...")
        >>>
        >>> # Build transaction
        >>> unsigned_tx = await adapter.build_transaction(EvmTxParams(
        ...     from_address="0x...",
        ...     to="0x...",
        ...     value="1000000000000000000",  # 1 ETH
        ... ))
    """

    def __init__(self, config: EvmChainConfig) -> None:
        self._config = config
        self._current_rpc_index = 0

    @property
    def chain_id(self) -> int:
        """Get chain ID."""
        return self._config.chain_id

    @property
    def symbol(self) -> str:
        """Get native currency symbol."""
        return self._config.symbol

    @property
    def decimals(self) -> int:
        """Get native currency decimals."""
        return self._config.decimals

    async def get_balance(self, address: str) -> Balance:
        """Get balance for an address."""
        result = await self._rpc_call("eth_getBalance", [address, "latest"])
        raw_value = int(result, 16)
        formatted = self._format_value(raw_value)

        return Balance(
            raw=str(raw_value),
            formatted=formatted,
            symbol=self._config.symbol,
            decimals=self._config.decimals,
        )

    async def get_nonce(self, address: str) -> int:
        """Get nonce for an address."""
        result = await self._rpc_call("eth_getTransactionCount", [address, "latest"])
        return int(result, 16)

    async def get_gas_prices(self) -> dict[str, int]:
        """Get current gas prices (EIP-1559)."""
        # Get latest block for base fee
        block = await self._rpc_call("eth_getBlockByNumber", ["latest", False])
        base_fee = int(block.get("baseFeePerGas", "0x0"), 16)

        # Get priority fee suggestion
        try:
            priority_fee = await self._rpc_call("eth_maxPriorityFeePerGas", [])
            max_priority_fee = int(priority_fee, 16)
        except Exception:
            max_priority_fee = 1_000_000_000  # Default to 1 gwei

        max_fee = base_fee * 2 + max_priority_fee

        return {
            "base_fee": base_fee,
            "max_fee_per_gas": max_fee,
            "max_priority_fee_per_gas": max_priority_fee,
        }

    async def estimate_gas(self, params: EvmTxParams) -> int:
        """Estimate gas for a transaction."""
        result = await self._rpc_call(
            "eth_estimateGas",
            [
                {
                    "from": params.from_address,
                    "to": params.to,
                    "value": hex(int(params.value)),
                    "data": params.data or "0x",
                }
            ],
        )
        return int(result, 16)

    async def build_transaction(self, params: EvmTxParams) -> UnsignedEvmTx:
        """Build an unsigned transaction."""
        import json
        import hashlib

        # Get nonce if not provided
        nonce = params.nonce
        if nonce is None:
            nonce = await self.get_nonce(params.from_address)

        # Get gas prices
        gas_prices = await self.get_gas_prices()
        max_fee = params.max_fee_per_gas or gas_prices["max_fee_per_gas"]
        max_priority_fee = params.max_priority_fee_per_gas or gas_prices["max_priority_fee_per_gas"]

        # Estimate gas if not provided
        gas_limit = params.gas_limit
        if gas_limit is None:
            estimated = await self.estimate_gas(params)
            gas_limit = int(estimated * 1.2)  # 20% buffer

        # Build transaction
        tx = {
            "chainId": self._config.chain_id,
            "nonce": nonce,
            "maxFeePerGas": max_fee,
            "maxPriorityFeePerGas": max_priority_fee,
            "gasLimit": gas_limit,
            "to": params.to,
            "value": int(params.value),
            "data": params.data or "",
        }

        # Serialize (simplified)
        serialized = json.dumps(tx).encode()

        # Compute signing hash (simplified)
        signing_hash = hashlib.sha256(serialized).digest()

        # Estimate fee
        estimated_fee = gas_limit * max_fee

        return UnsignedEvmTx(
            chain_id=self._config.chain_id,
            serialized=serialized,
            signing_hash=signing_hash,
            summary={
                "from": params.from_address,
                "to": params.to,
                "value": self._format_value(int(params.value)),
                "estimated_fee": self._format_value(estimated_fee),
            },
        )

    def finalize_transaction(
        self, unsigned_tx: UnsignedEvmTx, signature: Signature
    ) -> bytes:
        """Finalize a transaction with signature."""
        r = bytes.fromhex(signature.r.removeprefix("0x"))
        s = bytes.fromhex(signature.s.removeprefix("0x"))
        v = signature.recovery_id

        signed = unsigned_tx.serialized + r + s + bytes([v])
        return signed

    async def broadcast(self, signed_tx: bytes) -> TxHash:
        """Broadcast a signed transaction."""
        tx_hex = "0x" + signed_tx.hex()
        result = await self._rpc_call("eth_sendRawTransaction", [tx_hex])

        explorer_url = None
        if self._config.explorer_url:
            explorer_url = f"{self._config.explorer_url}/tx/{result}"

        return TxHash(hash=result, explorer_url=explorer_url)

    async def wait_for_confirmation(
        self, tx_hash: str, timeout_secs: int = 60
    ) -> TxReceipt:
        """Wait for transaction confirmation."""
        import time

        start_time = time.time()

        while time.time() - start_time < timeout_secs:
            try:
                receipt = await self._rpc_call("eth_getTransactionReceipt", [tx_hash])
                if receipt:
                    return TxReceipt(
                        tx_hash=tx_hash,
                        block_number=int(receipt["blockNumber"], 16),
                        status="success" if receipt["status"] == "0x1" else "failed",
                        gas_used=int(receipt["gasUsed"], 16),
                        effective_gas_price=int(receipt.get("effectiveGasPrice", "0x0"), 16),
                    )
            except Exception:
                pass

            await asyncio.sleep(2)

        raise TimeoutError(f"Transaction {tx_hash} not confirmed within timeout")

    def is_valid_address(self, address: str) -> bool:
        """Check if an address is valid."""
        import re
        return bool(re.match(r"^0x[0-9a-fA-F]{40}$", address))

    def get_explorer_tx_url(self, tx_hash: str) -> str | None:
        """Get explorer URL for a transaction."""
        if not self._config.explorer_url:
            return None
        return f"{self._config.explorer_url}/tx/{tx_hash}"

    def get_explorer_address_url(self, address: str) -> str | None:
        """Get explorer URL for an address."""
        if not self._config.explorer_url:
            return None
        return f"{self._config.explorer_url}/address/{address}"

    async def _rpc_call(self, method: str, params: list[Any]) -> Any:
        """Make an RPC call with failover."""
        errors: list[str] = []

        async with httpx.AsyncClient(timeout=30.0) as client:
            for _ in range(len(self._config.rpc_urls)):
                rpc_url = self._config.rpc_urls[self._current_rpc_index]

                try:
                    response = await client.post(
                        rpc_url,
                        json={
                            "jsonrpc": "2.0",
                            "method": method,
                            "params": params,
                            "id": 1,
                        },
                    )
                    data = response.json()

                    if "error" in data:
                        raise Exception(data["error"].get("message", "RPC error"))

                    return data["result"]

                except Exception as e:
                    errors.append(str(e))
                    self._current_rpc_index = (
                        self._current_rpc_index + 1
                    ) % len(self._config.rpc_urls)

        raise Exception(f"All RPC endpoints failed: {', '.join(errors)}")

    def _format_value(self, value: int) -> str:
        """Format a wei value."""
        divisor = 10 ** self._config.decimals
        whole = value // divisor
        fraction = value % divisor

        if fraction == 0:
            return f"{whole} {self._config.symbol}"

        fraction_str = str(fraction).zfill(self._config.decimals)
        trimmed = fraction_str.rstrip("0")
        return f"{whole}.{trimmed} {self._config.symbol}"
